### PR TITLE
Add a NoopEngine implementation

### PR DESCRIPTION
**Pull request by dakrone**
This adds a new Engine implementation that does.. nothing. Any operations throw
an `UnsupportedOperationException` when tried. This engine is intended as a
building block for replicated closed indices in subsequent code changes.

Relates to #31141

### CODE REVIEW

**Review by jasontedor — CHANGES_REQUESTED** [review]
I left some high-level comments. I have not done a careful review.

File: server/src/main/java/org/elasticsearch/index/engine/InternalEngineFactory.java [hunk 1]
```diff
@@ -24,4 +24,9 @@
     public Engine newReadWriteEngine(EngineConfig config) {
         return new InternalEngine(config);
     }
+
+    @Override
+    public Engine newNoopEngine(EngineConfig config) {
```
[the remark this answers is not in the document]
Comment (dakrone) [reply server/src/main/java/org/elasticsearch/index/engine/InternalEngineFactory.java]: Do we want to make this a plugin? I was thinking built-in but I guess we could move it to a module if we wanted to, I wasn't sure if it's a good idea to do engine implementations inside of modules

[the remark this answers is not in the document]
Comment (dakrone) [reply server/src/main/java/org/elasticsearch/index/engine/InternalEngineFactory.java]: Looking at this again, it appears that `EnginePlugin` is part of the CCR branch, are there plans to factor it out into a separate PR that goes into master?

[the remark this answers is not in the document]
Comment (jasontedor) [reply server/src/main/java/org/elasticsearch/index/engine/InternalEngineFactory.java]: I think it would be great if we could make it a module, I am always happy to see isolation when it makes sense and here I think it can. Note that CCR has an engine implementation inside a module. 😇

We have talked exactly about the possibility of reusing the engine plugin work in the context of closed indices many months ago so we could bring that work out of the CCR branch and into 6.x/master.

[the remark this answers is not in the document]
Comment (dakrone) [reply server/src/main/java/org/elasticsearch/index/engine/InternalEngineFactory.java]: Okay, I've removed the changes to the `EngineFactory` and I can use the pluggable one for the next PR (since this one doesn't hook the engine in yet)

File: server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java [hunk 2]
```diff
@@ -0,0 +1,334 @@
...
+    }
+
+    @Override
+    public void deactivateThrottling() {
+        throw new UnsupportedOperationException("closed engine can't throttle");
+    }
+
+    @Override
+    public void restoreLocalCheckpointFromTranslog() {
+
```
Comment (s1monw) [inline server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java]: extra NL

File: server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+    public NoopEngine(EngineConfig engineConfig) {
+        super(engineConfig);
+
+        store.incRef();
+        boolean success = false;
+        Translog translog = null;
+
+        // The deletion policy for the translog should not keep any translogs around, so the min age/size is set to -1
+        final TranslogDeletionPolicy translogDeletionPolicy = new TranslogDeletionPolicy(-1, -1);
```
Comment (s1monw) [inline server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java]: can we move this into the try block please?

File: server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
+ * Directory so that the last commit's user data can be read for the historyUUID
+ * and last committed segment info.
+ */
+public class NoopEngine extends Engine {
+
+    private final Translog translog;
+    private final IndexCommit lastCommit;
+    private final LocalCheckpointTracker localCheckpointTracker;
+    private final String historyUUID;
+    private SegmentInfos lastCommittedSegmentInfos;
```
Comment (s1monw) [inline server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java]: this can be final no?

File: server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java [hunk 5]
```diff
@@ -0,0 +1,334 @@
...
+ * required in order to have an engine. All attempts to do something (search,
+ * index, get), throw {@link UnsupportedOperationException}. This does maintain
+ * a translog with a deletion policy so that when flushing, no translog is
+ * retained on disk (setting a retention size and age of 0).
+ *
+ * It's also important to notice that this does list the commits of the Store's
+ * Directory so that the last commit's user data can be read for the historyUUID
+ * and last committed segment info.
+ */
+public class NoopEngine extends Engine {
```
Comment (s1monw) [inline server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java]: can this class be final and maybe pkg private?

File: server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java [hunk 6]
```diff
@@ -0,0 +1,334 @@
...
+
+    @Override
+    public CommitId flush(boolean force, boolean waitIfOngoing) throws EngineException {
+        return new CommitId(lastCommittedSegmentInfos.getId());
+    }
+
+    @Override
+    public CommitId flush() throws EngineException {
+        try {
+            translog.rollGeneration();
```
Comment (s1monw) [inline server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java]: this looks wrong. We don't write anything here why do we need to modify the translog? I think this should be read-only

**Review by s1monw — CHANGES_REQUESTED** [review]
left some comments.

File: server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java (hunk 6 above)
Comment (dakrone) [reply server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java]: In order for flushing to clear existing translogs (because we don't want any translog operations to be replayed for peer or store recovery) we want the flush method to remove the translog, this was added so that flushing the new engine would ensure that we don't have any translog operations around that could cause UOEs during recovery

File: server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java (hunk 5 above)
Comment (dakrone) [reply server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java]: Sure, though I'll un-final it when I work on the second half of this, but it can be final for now :)

File: server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java (hunk 4 above)
Comment (dakrone) [reply server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java]: Yep

File: server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java (hunk 3 above)
Comment (dakrone) [reply server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java]: Certainly

File: server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java (hunk 6 above)
Comment (dakrone) [reply server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java]: I can remove this for now and re-introduce it later when adding the state transition part, if that makes it better, but we still need to be able to completely remove translog ops before doing recovery since we have no way to do operation-based recovery.

What do you think?

Comment (s1monw) [reply server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java]: yeah I'd like to remove it for now I can 't see in this change why it's needed

Comment (dakrone) [reply server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java]: Okay, I've removed that change from this PR

File: server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java [hunk 7]
```diff
@@ -0,0 +1,326 @@
...
+            this.translog = translog;
+            List<IndexCommit> indexCommits = DirectoryReader.listCommits(store.directory());
+            lastCommit = indexCommits.get(indexCommits.size()-1);
+            historyUUID = lastCommit.getUserData().get(HISTORY_UUID_KEY);
+            // We don't want any translogs hanging around for recovery, so we need to set these accordingly
+            final long lastGen = Long.parseLong(lastCommit.getUserData().get(Translog.TRANSLOG_GENERATION_KEY));
+            translogDeletionPolicy.setTranslogGenerationOfLastCommit(lastGen);
+            translogDeletionPolicy.setMinTranslogGenerationForRecovery(lastGen);
+
+            localCheckpointTracker = createLocalCheckpointTracker();
```
Comment (bleskes) [inline server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java]: we don't really need a local checkpoint tracker here, instead can we work on master to not expose the localCheckpointTracker out of engine (similar to how we don't expose the translog) and then we can avoid creating it? We should assert that maxSeq == localCheckpoint when opening lucene and otherwise fail.

File: server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java [hunk 8]
```diff
@@ -0,0 +1,326 @@
...
+        store.incRef();
+        boolean success = false;
+        Translog translog = null;
+
+        try {
+            // The deletion policy for the translog should not keep any translogs around, so the min age/size is set to -1
+            final TranslogDeletionPolicy translogDeletionPolicy = new TranslogDeletionPolicy(-1, -1);
+
+            lastCommittedSegmentInfos = store.readLastCommittedSegmentsInfo();
+            translog = openTranslog(engineConfig, translogDeletionPolicy, engineConfig.getGlobalCheckpointSupplier());
```
Comment (bleskes) [inline server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java]: Do we really need a translog? all we want is to validate that the translog has a the right uuid and that it's empty?

File: server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java [hunk 9]
```diff
@@ -0,0 +1,326 @@
...
+    }
+
+    @Override
+    public Translog getTranslog() {
+        return translog;
+    }
+
+    @Override
+    public boolean ensureTranslogSynced(Stream<Translog.Location> locations) {
+        return false;
```
Comment (bleskes) [inline server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java]: I think this can throw an unsupported operation exception too?

**Review by bleskes — CHANGES_REQUESTED** [review]
Thanks @dakrone . I left some comments and questions.

File: server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java (hunk 7 above)
Comment (dakrone) [reply server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java]: Do you mean something like this: https://github.com/elastic/elasticsearch/pull/31213

Comment (bleskes) [reply server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java]: yes. thanks.

File: server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java (hunk 8 above)
Comment (dakrone) [reply server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java]: We use the translog all over the place in other methods in `Engine`. We can try to encapsulate all of these into different methods (similar to #31213), but I'm not sure what benefit we'd actually get from that. In the next iteration we'll need the translog because we'll need to sync and flush it on engine opening so that there are no operations to be replayed during peer recovery (which I think will still need a translog to retrieve stats about # of ops).

File: server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java [hunk 10]
```diff
@@ -0,0 +1,326 @@
...
+        throw new UnsupportedOperationException("gets are not supported on a noop engine");
+    }
+
+    @Override
+    public Searcher acquireSearcher(String source, SearcherScope scope) throws EngineException {
+        throw new UnsupportedOperationException("searching is not supported on a noop engine");
+    }
+
+    @Override
+    public Translog getTranslog() {
```
Comment (bleskes) [inline server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java]: this increases visibilty

File: server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java (hunk 8 above)
Comment (bleskes) [reply server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java]: > I'm not sure what benefit we'd actually get from that

I looked at these methods and I agree it's on the edge - there are quite a few. I would still prefer we replace them with unsupported operations or noop returns (ex empty sanpshots). This will lock things down and prevent things that aren't supposed to happen - I think that's good. An alternative is to implement a NoopTranslog but that's another rabbit hole.

> we'll need to sync and flush it on engine opening

Why is that? I think it's good to only close indices that have no ongoing indexing (like our plan for frozen index). Regardless - why can't we do the flush / trim when we close the open engine and convert it to a noop engine? 

I can see one thing down the road because we may close an index on recovery where it has broken settings (TBD). In that case I would still prefer to make utilities methods like `Store#associateIndexWithNewTranslog` that work on the folder. Note that this problem isn't solved even if we keep the translog open as we can't index operations from it into the lucene index with the `NoopEngine` nor ship to a recovering shard with `NoopEngine` in it. We assume those don't exist. I think we should discuss this separately.

Comment (dakrone) [reply server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java]: > I think it's good to only close indices that have no ongoing indexing (like our plan for frozen index).

Yes, but even with no ongoing indexing, a translog still remains (due to retention policy)

> Regardless - why can't we do the flush / trim when we close the open engine and convert it to a noop engine?

That would require setting a new retention policy on an existing engine (making a part of InternalEngine mutable which makes me `:(`). In the future though, we could do the retention policy, sync, and flush/trim when the NoopEngine is opened, and then immediately close the translog.

In order to do this though, we'll have to remove the `getTranslog` method from Engine, is that something you want me to do as a precursor to this?

**Comment by dakrone** [comment]
@jason @bleskes @s1monw I've removed almost all the pieces from this that were deemed unnecessary, can you take another look please?

File: server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java [hunk 11]
```diff
@@ -0,0 +1,343 @@
...
+        store.incRef();
+        boolean success = false;
+
+        try {
+            lastCommittedSegmentInfos = store.readLastCommittedSegmentsInfo();
+            List<IndexCommit> indexCommits = DirectoryReader.listCommits(store.directory());
+            lastCommit = indexCommits.get(indexCommits.size()-1);
+            historyUUID = lastCommit.getUserData().get(HISTORY_UUID_KEY);
+            // We don't want any translogs hanging around for recovery, so we need to set these accordingly
+            final long lastGen = Long.parseLong(lastCommit.getUserData().get(Translog.TRANSLOG_GENERATION_KEY));
```
Comment (bleskes) [inline server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java]: this is not used?

Also - I think it's to validate integrity - i.e. open the translog, see that it's uuid matches, see that it's empty and shut it down?

File: server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java [hunk 12]
```diff
@@ -0,0 +1,343 @@
...
+    public void waitForOpsToComplete(long seqNo) {
+    }
+
+    @Override
+    public void resetLocalCheckpoint(long localCheckpoint) {
+    }
+
+    @Override
+    public SeqNoStats getSeqNoStats(long globalCheckpoint) {
+        return new SeqNoStats(0, 0, 0);
```
Comment (bleskes) [inline server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java]: use the incoming globalCheckpoint?

**Review by bleskes — CHANGES_REQUESTED** [review]
Looks great thanks for the extra iteration. I left some minor asks.

File: server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java (hunk 12 above)
Comment (dakrone) [reply server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java]: Wouldn't that violate the assertion that global checkpoint is the minimum of all shards' local checkpoints?

File: server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java (hunk 11 above)
Comment (dakrone) [reply server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java]: Okay I pushed a commit removing the unused `lastGen` and open-closing the translog for validation purposes

File: server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java [hunk 13]
```diff
@@ -0,0 +1,374 @@
...
+            lastCommittedSegmentInfos = store.readLastCommittedSegmentsInfo();
+            List<IndexCommit> indexCommits = DirectoryReader.listCommits(store.directory());
+            lastCommit = indexCommits.get(indexCommits.size()-1);
+            historyUUID = lastCommit.getUserData().get(HISTORY_UUID_KEY);
+
+            // The deletion policy for the translog should not keep any translogs around, so the min age/size is set to -1
+            final TranslogDeletionPolicy translogDeletionPolicy = new TranslogDeletionPolicy(-1, -1);
+
+            // The translog is opened and closed to validate that the translog UUID from lucene is the same as the one in the translog
+            Translog translog = openTranslog(engineConfig, translogDeletionPolicy, engineConfig.getGlobalCheckpointSupplier());
```
Comment (bleskes) [inline server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java]: can you check the translog is empty?

File: server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java (hunk 12 above)
Comment (bleskes) [reply server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java]: that would but we don't check this here. Also - I think we should load the stats from the last commit point, not just 0 (both max and local checkpoint are stored in the commit)

File: server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java [hunk 14]
```diff
@@ -0,0 +1,374 @@
...
+    public long getLocalCheckpoint() {
+        return 0;
+    }
+
+    @Override
+    public void waitForOpsToComplete(long seqNo) {
+    }
+
+    @Override
+    public void resetLocalCheckpoint(long localCheckpoint) {
```
Comment (bleskes) [inline server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java]: can you assert this is equal to getLocalCheckpoint()?

File: server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java [hunk 15]
```diff
@@ -0,0 +1,374 @@
...
+        return EMPTY_TRANSLOG_STATS;
+    }
+
+    @Override
+    public Translog.Location getTranslogLastWriteLocation() {
+        return EMPTY_TRANSLOG_LOCATION;
+    }
+
+    @Override
+    public long getLocalCheckpoint() {
```
Comment (bleskes) [inline server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java]: can you return the local checkpoint stored in the commit?

File: server/src/test/java/org/elasticsearch/index/engine/NoopEngineTests.java [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+import org.elasticsearch.index.IndexSettings;
+import org.elasticsearch.test.IndexSettingsModule;
+
+import java.io.IOException;
+
+import static org.hamcrest.Matchers.equalTo;
+
+public class NoopEngineTests extends EngineTestCase {
+    private static final IndexSettings INDEX_SETTINGS = IndexSettingsModule.newIndexSettings("index", Settings.EMPTY);
+
```
Comment (bleskes) [inline server/src/test/java/org/elasticsearch/index/engine/NoopEngineTests.java]: can you also add a test that creates a normal engine, adds some data, flushes, closes and the opens a noop engine and verify all the stats makes sense and getting a lucene snapshot returns the expected data (doc counts etc)?

**Review by bleskes — CHANGES_REQUESTED** [review]
I did another round. Sorry for not thinking about these in the first run.

File: server/src/test/java/org/elasticsearch/index/engine/NoopEngineTests.java (hunk 16 above)
Comment (dakrone) [reply server/src/test/java/org/elasticsearch/index/engine/NoopEngineTests.java]: > getting a lucene snapshot returns the expected data (doc counts etc)?

Do you mean from `newTranslogSnapshotFromMinSeqNo`? I've been returning an empty snapshot for this with docCount of 0 (and no operations, because we have no way of returning operations), so it won't match the same doc count that a snapshot from a regular engine will.

Comment (bleskes) [reply server/src/test/java/org/elasticsearch/index/engine/NoopEngineTests.java]: I meant [this](https://github.com/elastic/elasticsearch/pull/31163/files#diff-2ec2c2b070f96bf66b888db94648ffe0R321) . The standard engine ends up returning a `SnapshotIndexCommit` which is why I used the term lucene snapshot. I hope this is clearer.

File: server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java (hunk 13 above)
Comment (dakrone) [reply server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java]: Sure I added this and the test you requested

File: server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java (hunk 14 above)
Comment (dakrone) [reply server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java]: Yep added this

File: server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java (hunk 15 above)
Comment (dakrone) [reply server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java]: Sure, added this

File: server/src/test/java/org/elasticsearch/index/engine/NoopEngineTests.java (hunk 16 above)
Comment (dakrone) [reply server/src/test/java/org/elasticsearch/index/engine/NoopEngineTests.java]: Yep, thanks for clarifying, this was added as a test

**Comment by dakrone** [comment]
@bleskes thanks for taking another look, I think I addressed all your feedback

File: server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java [hunk 17]
```diff
@@ -98,6 +98,9 @@ public void close() {
 
             // The translog is opened and closed to validate that the translog UUID from lucene is the same as the one in the translog
             Translog translog = openTranslog(engineConfig, translogDeletionPolicy, engineConfig.getGlobalCheckpointSupplier());
```
Comment (bleskes) [inline server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java]: wrap this in try finally?

File: server/src/test/java/org/elasticsearch/index/engine/NoopEngineTests.java [hunk 18]
```diff
@@ -56,4 +67,38 @@ public void testTwoNoopEngines() throws IOException {
         engine2.close();
     }
 
+    public void testNoopAfterRegularEngine() throws IOException {
+        int docs = randomIntBetween(1, 10);
+        ReplicationTracker tracker = (ReplicationTracker) engine.config().getGlobalCheckpointSupplier();
```
Comment (bleskes) [inline server/src/test/java/org/elasticsearch/index/engine/NoopEngineTests.java]: why is this needed? we typically just write to the engine?

File: server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java [hunk 19]
```diff
@@ -0,0 +1,384 @@
...
+
+        store.incRef();
+        boolean success = false;
+
+        try {
+            lastCommittedSegmentInfos = store.readLastCommittedSegmentsInfo();
+            List<IndexCommit> indexCommits = DirectoryReader.listCommits(store.directory());
+            lastCommit = indexCommits.get(indexCommits.size()-1);
+            historyUUID = lastCommit.getUserData().get(HISTORY_UUID_KEY);
+            localCheckpoint = Long.parseLong(lastCommit.getUserData().get(SequenceNumbers.LOCAL_CHECKPOINT_KEY));
```
Comment (bleskes) [inline server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java]: This is not yet relevant to this PR but these will have to be the same IMO .

**Review by bleskes — APPROVED** [review]
LGTM. Thanks for the extra work @dakrone . I left some question on the test, but I'm happy either way.

File: server/src/test/java/org/elasticsearch/index/engine/NoopEngineTests.java (hunk 18 above)
Comment (dakrone) [reply server/src/test/java/org/elasticsearch/index/engine/NoopEngineTests.java]: This is a rabbit hole @jasontedor and I got into last night; in order to completely flush and remove the translog we need a deletion policy where we can advance the generation needed for recovery, in order to advance that we have to have the tracker working correctly (which I guess is normally managed at the IndexShard level), so this allows us to advance global checkpoints as needed so we can remove the translog and open a new noop engine (otherwise the noop engine would throw an exception about there being translog operations when being opened).

File: server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java (hunk 17 above)
Comment (dakrone) [reply server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java]: Yes good catch, I thought about that last night but forgot to do it, I'll update that

File: server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java [hunk 20]
```diff
@@ -0,0 +1,384 @@
...
+    NoopEngine(EngineConfig engineConfig) {
+        super(engineConfig);
+
+        store.incRef();
+        boolean success = false;
+
+        try {
+            lastCommittedSegmentInfos = store.readLastCommittedSegmentsInfo();
+            List<IndexCommit> indexCommits = DirectoryReader.listCommits(store.directory());
+            lastCommit = indexCommits.get(indexCommits.size()-1);
```
Comment (jasontedor) [inline server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java]: Formatting nit: `indexCommits.size()-1` -> `indexCommits.size() - 1`

File: server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java [hunk 21]
```diff
@@ -0,0 +1,384 @@
...
+            historyUUID = lastCommit.getUserData().get(HISTORY_UUID_KEY);
+            localCheckpoint = Long.parseLong(lastCommit.getUserData().get(SequenceNumbers.LOCAL_CHECKPOINT_KEY));
+            maxSeqNo = Long.parseLong(lastCommit.getUserData().get(SequenceNumbers.MAX_SEQ_NO));
+
+            // The deletion policy for the translog should not keep any translogs around, so the min age/size is set to -1
+            final TranslogDeletionPolicy translogDeletionPolicy = new TranslogDeletionPolicy(-1, -1);
+
+            // The translog is opened and closed to validate that the translog UUID from lucene is the same as the one in the translog
```
Comment (jasontedor) [inline server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java]: Can we have a test case for this?

File: server/src/test/java/org/elasticsearch/index/engine/NoopEngineTests.java [hunk 22]
```diff
@@ -0,0 +1,105 @@
...
+
+public class NoopEngineTests extends EngineTestCase {
+    private static final IndexSettings INDEX_SETTINGS = IndexSettingsModule.newIndexSettings("index", Settings.EMPTY);
+
+    public void testNoopEngine() throws IOException {
+        engine.close();
+        final NoopEngine engine = new NoopEngine(noopConfig(INDEX_SETTINGS, store, primaryTranslogDir));
+        expectThrows(UnsupportedOperationException.class, () -> engine.index(null));
+        expectThrows(UnsupportedOperationException.class, () -> engine.delete(null));
+        expectThrows(UnsupportedOperationException.class, () -> engine.noOp(null));
```
Comment (jasontedor) [inline server/src/test/java/org/elasticsearch/index/engine/NoopEngineTests.java]: As a follow-up can we have:
 - `Noop` -> `NoOp`
 - `noop` -> `no-op` (when written in exception messages to the user, and code comments)
 - `noopEngine` -> `noOpEngine

File: server/src/test/java/org/elasticsearch/index/engine/NoopEngineTests.java [hunk 23]
```diff
@@ -0,0 +1,105 @@
...
+        expectThrows(UnsupportedOperationException.class, engine::activateThrottling);
+        expectThrows(UnsupportedOperationException.class, engine::deactivateThrottling);
+        assertThat(engine.refreshNeeded(), equalTo(false));
+        assertThat(engine.shouldPeriodicallyFlush(), equalTo(false));
+        engine.close();
+    }
+
+    public void testTwoNoopEngines() throws IOException {
+        engine.close();
+        // It's so noop you can even open two engines for the same store without tripping anything
```
Comment (jasontedor) [inline server/src/test/java/org/elasticsearch/index/engine/NoopEngineTests.java]: Is this something that we are going to rely on?

**Review by jasontedor — COMMENTED** [review]
I left some comments/questions.

**Review by s1monw — APPROVED** [review]

File: server/src/test/java/org/elasticsearch/index/engine/NoopEngineTests.java (hunk 23 above)
Comment (dakrone) [reply server/src/test/java/org/elasticsearch/index/engine/NoopEngineTests.java]: Probably not? I just thought it was a good test that we aren't acquiring locks or locking files accidentally or anything like that

File: server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java (hunk 21 above)
Comment (dakrone) [reply server/src/main/java/org/elasticsearch/index/engine/NoopEngine.java]: Sure, added a test

File: server/src/test/java/org/elasticsearch/index/engine/NoopEngineTests.java (hunk 22 above)
Comment (dakrone) [reply server/src/test/java/org/elasticsearch/index/engine/NoopEngineTests.java]: ooookkaayy

**Comment by dakrone** [comment]
@jasontedor I addressed your comments (I'll do the rename in a followup)

File: server/src/test/java/org/elasticsearch/index/engine/NoopEngineTests.java (hunk 23 above)
Comment (jasontedor) [reply server/src/test/java/org/elasticsearch/index/engine/NoopEngineTests.java]: Okay, maybe the test name should be clearer, or at least a comment of its purpose?

**Review by jasontedor — APPROVED** [review]
LGTM.

File: server/src/test/java/org/elasticsearch/index/engine/NoopEngineTests.java (hunk 23 above)
Comment (dakrone) [reply server/src/test/java/org/elasticsearch/index/engine/NoopEngineTests.java]: Sure I'll add a comment

**Comment by dakrone** [comment]
@elasticmachine run sample packaging tests please
